backend/corpora/common/dataset_validator.py

```diff
@@ -0,0 +1,119 @@
+        self.s3_file_system = s3fs.S3FileSystem()
+
+    def validate_dataset_file(self):
+        """
+        Reads in file object and triages for specific file type validation.
+        """
+
+        file_object = self.s3_file_system.open(self.s3_path, "rb")
+        file_object_size = file_object.info().get("Size")
+        logging.info(f"Validating file {self.s3_uri} with size {sizeof_formatted(file_object_size)}")
+
+        if self.s3_path.endswith(CorporaConstants.H5AD_FILE_TYPE):
+            self.validate_h5ad_dataset(file_object)
+
+        elif self.s3_path.endswith(CorporaConstants.LOOM_FILE_TYPE):
+            self.validate_loom_dataset(file_object)
+
+        else:
+            logging.warning(f"Unknown type of dataset with path {self.s3_path}!")
+
+        file_object.close()
+
+    def validate_h5ad_dataset(self, file_object):
+        """
+        Reads the file contents into an AnnData object. Each attribute of the AnnData object will then be checked to
+        ensure it contains the appropriate metadata.
+        """
+
+        start_time = time.time()
+        logging.info("Reading anndata object...")
+        anndata_object = anndata.read_h5ad(file_object)
+        logging.info(f"Finished reading anndata object in {time.time() - start_time:.3f} seconds.")
+
+        start_time = time.time()
+        logging.info("Beginning validation of file...")
+        self.verify_obs(anndata_object)
+        self.verify_vars(anndata_object)
+        self.verify_uns(anndata_object)
+        logging.info(f"Finished completing validation on the file in {time.time() - start_time:.3f} seconds.")
+
+    def verify_obs(self, data_object: anndata.AnnData):
+        """
+        Validates the observation attribute of an AnnData object. Checks to ensure that all observation IDs are
+        unique and that the observation metadata fields as described by the Corpora Schema exist. If the validation
+        fails in any way, the errors are outputted rather than the validation aborted.
+        """
+
+        # Check to ensure that all IDs are unique
+        observation_ids = data_object.obs.index.tolist()
+        if len(observation_ids) != len(set(observation_ids)):
+            logging.warning("Each observation is not unique!")
+
+        obs_keys = map(str.upper, data_object.obs_keys())
+        for metadata_field in (
+            CorporaConstants.REQUIRED_OBSERVATION_METADATA_FIELDS
+            + CorporaConstants.REQUIRED_OBSERVATION_ONTOLOGY_METADATA_FIELDS
+        ):
+            if metadata_field.upper() not in obs_keys:
+                self.log_error_message(metadata_field, "obs", type(data_object).__name__)
+
+    def verify_vars(self, data_object: anndata.AnnData):
+        """
+        Validates the variable attribute of the AnnData object to ensure that all variable IDs are unique.
+        """
+
+        variable_ids = data_object.var.index.tolist()
+
+        if len(variable_ids) != len(set(variable_ids)):
```

Comment:
I suppose this is where we want to make sure we don't have any "bad" gene names, but maybe that makes sense to break into a separate ticket?